### DISCUSSION TITLE
Node – Corepack missing? 🕵🏻

### DISCUSSION THREAD

**Discussion by blasterbug** [Homebrew/discussions#2348] [Casks] [answered]
Hello,

Node.js comes with a few utilities such as [corepack](https://nodejs.org/api/corepack.html). But it seems when installing node via homebrew those binaries are not available? 

I have node@17.0.1 installed and the corepack utility is missing.

**Comment by carlocab** [comment]
The docs say this is experimental, so I'm not sure we want to enable this feature.

However, if you think many users would find it useful, you're welcome to make a pull request against the Node formula to enable it. We don't disable it anywhere ([here](https://github.com/Homebrew/homebrew-core/blob/4f735e0aa222daf862a68d0ee42442518e7445a7/Formula/node.rb#L78-L100) are the options we pass), so it's likely just not included by default when building from source.

**Comment by rinsuki** [comment]
we are currently passing `--without-npm` to configure, and I guess `--without-npm` contains `--without-corepack` (or something else).

then we may need to one of these:

- install corepack manually in node formula, like npm does
- or provide as separate formula (like yarn?)
- or include node's npm instead of manually one and remove `--without-npm`.

**Reply by carlocab** [reply]
> * install corepack manually in node formula, like npm does

This is fine.

> * or provide as separate formula (like yarn?)

This is probably fine too, depending on how the separate formula looks.

> * or include node's npm instead of manually one and remove `--without-npm`.

This is probably not fine.

**Reply by rinsuki** [reply]
corepack says /usr/local/bin/{yarn,pnpm} are conflicts with corepack, then I think "provide corepack as separate formula" (and mark yarn and pnpm are conflicts with corepack in formula) is best choice.

if it is good, i going to create corepack formula.

do you think about this? @carlocab

**Reply by carlocab** [reply]
Sure; please open a pull request for review.

**Reply by rinsuki** [reply]
Created! https://github.com/Homebrew/homebrew-core/pull/88700

**Comment by kud** [comment]
Hey guys,

Sorry, I don't get it. I used to install `node` and not `corepack`, and now I always get that when I try to install/upgrade `node`.

```
==> Installing node
==> Pouring node--23.2.0.arm64_sequoia.bottle.tar.gz
Error: The `brew link` step did not complete successfully
The formula built, but is not symlinked into /opt/homebrew
Could not symlink bin/corepack
Target /opt/homebrew/bin/corepack
already exists. You may want to remove it:
  rm '/opt/homebrew/bin/corepack'

To force the link and overwrite all conflicting files:
  brew link --overwrite node

To list all files that would be deleted:
  brew link --overwrite node --dry-run

Possible conflicting files are:
/opt/homebrew/bin/corepack -> /opt/homebrew/lib/node_modules/corepack/dist/corepack.js
/opt/homebrew/lib/node_modules/corepack/CHANGELOG.md
/opt/homebrew/lib/node_modules/corepack/LICENSE.md
/opt/homebrew/lib/node_modules/corepack/README.md
/opt/homebrew/lib/node_modules/corepack/dist/corepack.js
/opt/homebrew/lib/node_modules/corepack/dist/lib/corepack.cjs
/opt/homebrew/lib/node_modules/corepack/dist/npm.js
/opt/homebrew/lib/node_modules/corepack/dist/npx.js
/opt/homebrew/lib/node_modules/corepack/dist/pnpm.js
/opt/homebrew/lib/node_modules/corepack/dist/pnpx.js
/opt/homebrew/lib/node_modules/corepack/dist/yarn.js
/opt/homebrew/lib/node_modules/corepack/dist/yarnpkg.js
/opt/homebrew/lib/node_modules/corepack/package.json
/opt/homebrew/lib/node_modules/corepack/shims/corepack
/opt/homebrew/lib/node_modules/corepack/shims/corepack.cmd
/opt/homebrew/lib/node_modules/corepack/shims/corepack.ps1
/opt/homebrew/lib/node_modules/corepack/shims/nodewin/corepack
/opt/homebrew/lib/node_modules/corepack/shims/nodewin/corepack.cmd
/opt/homebrew/lib/node_modules/corepack/shims/nodewin/corepack.ps1
/opt/homebrew/lib/node_modules/corepack/shims/nodewin/npm
/opt/homebrew/lib/node_modules/corepack/shims/nodewin/npm.cmd
/opt/homebrew/lib/node_modules/corepack/shims/nodewin/npm.ps1
/opt/homebrew/lib/node_modules/corepack/shims/nodewin/npx
/opt/homebrew/lib/node_modules/corepack/shims/nodewin/npx.cmd
/opt/homebrew/lib/node_modules/corepack/shims/nodewin/npx.ps1
/opt/homebrew/lib/node_modules/corepack/shims/nodewin/pnpm
/opt/homebrew/lib/node_modules/corepack/shims/nodewin/pnpm.cmd
/opt/homebrew/lib/node_modules/corepack/shims/nodewin/pnpm.ps1
/opt/homebrew/lib/node_modules/corepack/shims/nodewin/pnpx
/opt/homebrew/lib/node_modules/corepack/shims/nodewin/pnpx.cmd
/opt/homebrew/lib/node_modules/corepack/shims/nodewin/pnpx.ps1
/opt/homebrew/lib/node_modules/corepack/shims/nodewin/yarn
/opt/homebrew/lib/node_modules/corepack/shims/nodewin/yarn.cmd
/opt/homebrew/lib/node_modules/corepack/shims/nodewin/yarn.ps1
/opt/homebrew/lib/node_modules/corepack/shims/nodewin/yarnpkg
/opt/homebrew/lib/node_modules/corepack/shims/nodewin/yarnpkg.cmd
/opt/homebrew/lib/node_modules/corepack/shims/nodewin/yarnpkg.ps1
/opt/homebrew/lib/node_modules/corepack/shims/npm
/opt/homebrew/lib/node_modules/corepack/shims/npm.cmd
/opt/homebrew/lib/node_modules/corepack/shims/npm.ps1
/opt/homebrew/lib/node_modules/corepack/shims/npx
/opt/homebrew/lib/node_modules/corepack/shims/npx.cmd
/opt/homebrew/lib/node_modules/corepack/shims/npx.ps1
/opt/homebrew/lib/node_modules/corepack/shims/pnpm
/opt/homebrew/lib/node_modules/corepack/shims/pnpm.cmd
/opt/homebrew/lib/node_modules/corepack/shims/pnpm.ps1
/opt/homebrew/lib/node_modules/corepack/shims/pnpx
/opt/homebrew/lib/node_modules/corepack/shims/pnpx.cmd
/opt/homebrew/lib/node_modules/corepack/shims/pnpx.ps1
/opt/homebrew/lib/node_modules/corepack/shims/yarn
/opt/homebrew/lib/node_modules/corepack/shims/yarn.cmd
/opt/homebrew/lib/node_modules/corepack/shims/yarn.ps1
/opt/homebrew/lib/node_modules/corepack/shims/yarnpkg
/opt/homebrew/lib/node_modules/corepack/shims/yarnpkg.cmd
/opt/homebrew/lib/node_modules/corepack/shims/yarnpkg.ps1
```

Any idea how to resolve that?

Thank you

**Reply by SMillerDev** [reply]
From the output, you have 2 options. Either you remove the existing corepack or you force brew to overwrite the existing corepack.
```
You may want to remove it:
  rm '/opt/homebrew/bin/corepack'

To force the link and overwrite all conflicting files:
  brew link --overwrite node

To list all files that would be deleted:
  brew link --overwrite node --dry-run
```

**Reply by kud** [reply]
Yes, I actually did both, removing Corepack and also installing/linking Node.

Though, this issue comes back often, and I don't even install Corepack myself. Does it come from another package I install as a dependency, maybe?

Thanks by the way for helping me 😌

**Comment by thomaswitt** [comment]
Same problem here, similar discussion here https://github.com/nodejs/node/issues/56519

**Reply by botantony** [reply, marked as answer]
In October of the last year `corepack` formula was merged into `node`: https://github.com/Homebrew/homebrew-core/pull/195770
Merging packages is not very common thing (usually you do the opposite, split one big package into a few smaller ones), so unfortunately Homebrew does not have mechanisms to merge two formulae seamlessly. You had to either remove `corepack` binary manually (`rm '$(brew --prefix)/bin/corepack'`) or let Homebrew overwrite conflicting files (`brew link --overwrite node`)

**Reply by chenrui333** [reply]
close via https://github.com/Homebrew/homebrew-core/pull/195770
